Whisky/View Models/BottleVM.swift

```diff
@@ -59,4 +59,15 @@ class BottleVM: ObservableObject {
             }
         }
     }
+
+    func validateBottleName(bottleName: String) -> (isInvalidName: Bool, description: String) {
```

Comment:
I would strongly recommend not returning a tuple here.
Instead, consider an enum with associated values:

```swift
enum BottleValidationResult {
  case success
  case failure(reason: FailureReason)

  enum FailureReason {
    case emptyName
    case alreadyExists
  }
}
```

As you can see from your `(false, "")` case, with a tuple here, you *have* to return a `String`, even if it's an empty `String`. It's also confusing to see `false` as a success value, IMO.

Moving to an enum like this removes the need to attach a `String` in the success case, and using an enum for the failure reason means that the calling code can know exactly why it failed without directly inspecting or comparing strings.

Reply:
I initially fought about doing it this way, as it's the more correct way to do it, but decided not to, as for such a simple validation it would be a bit of an overkill, and would require more logic in the View. 

This would be probably a lot more useful for when the bottle creation fails, as then you can propagate up the error. It then would be used to decide how to perform the cleanup, and show the correct error notification.

Reply:
I disagree strongly that it is overkill to model types correctly.

It's safer, and it's much easier for everyone to deal with. It's no real extra effort - you simply switch on the result instead of directly assigning the tuple values to state properties.

Reply:
I disagree strongly that it is overkill to model types correctly.

It's safer, and it's much easier for everyone to deal with. It's no real extra effort - you simply `switch` on the result instead of directly assigning the tuple values to state properties.

You can add a small extension for the sake of getting a `String` to display on the UI.

Reply:
Ok, I rewrote it using enums